### ISSUE TITLE
Implementing Fourier Features into NERF

### ISSUE THREAD

**Issue by phongnhhn92** [kwea123/nerf_pl#24]
Hey, I think you have noticed this work https://github.com/tancik/fourier-feature-networks. Do you think it is straight forward to replace the positional encoding with this Fourier Features. I have tried and didnt see much improvement.

**Comment by kwea123** [comment]
Yes, I am currently experimenting. I have discussed with someone here https://github.com/bmild/nerf/issues/60
What I found is different than the paper as well... I found gaussian features to be worse, then I also tried uniform sampling and found it better when using more features, but in the end sampling random numbers is not necessary at all, just use the current linspace with more points suffices...
https://github.com/kwea123/nerf_pl/blob/34bb7a9c24a6475a1cb711007a8efc4634129175/models/nerf.py#L17

To conclude I tried the followings embeddings:
1. sin(ax+by+cz) with a,b,c ~ N(0, var)
2. sin(ax+by+cz) with a,b,c ~ U(0, max)
3. sin(ax), sin(by), sin(cz) with a, b, c ~ N(0, var)
4. sin(ax), sin(by), sin(cz) with a, b, c ~ U(0, max)
5. sin(ax), sin(ay), sin(az) with a ~ U(0, max)
6. sin(ax), sin(ay), sin(az) with `a` fixed beforehand (the current PE)

And it turns out 6. is the best, and it gets better if I use more frequencies (current PE uses 10, I increased it to 64 and it's better with +1 point PSNR)
the moidification just looks like this, very simple...
`self.freq_bands = 2**torch.linspace(0, N_freqs-1, 64)`

I didn't (and probably won't) perform any further analysis like drawing the NTK, so to be honest I don't know what's wrong with the settings, but there are some points that I think might be the problem:
1.  The experiment in the original paper is oversimplified: there is neither coarse to fine nor direction embedding. When combined with everything, the performance gain might not be that significant, or even get worse.
2.  The scene I use is a forward-facing scene which uses NDC, so it means the scale in x,y,z is different especially for z. Therefore it might require more careful tuning on the frequencies, notably we shouldn't be using the same distribution for all x,y,z.
Hope these insights help.
What scene and parameters did you try?

**Comment by phongnhhn92** [comment]
So i have followed the other thread as well and I found that adding Gaussian embedding does improve the result just a little bit but we if look at the result closely then there are small artifacts. I tested on my 2 scenes (I captured using my phone). The indoor scene work best with 64 frequencies, the another indoor looks good with 32. In the paper, they said that learning this Gaussian embedding doesn't help but I think that hand-tuning like this is sub-optimal.
